File: app/src/main/java/org/metabrainz/mobile/api/webservice/LookupService.java
package org.metabrainz.mobile.api.webservice;

import org.metabrainz.mobile.api.data.ArtistWikiSummary;
import org.metabrainz.mobile.api.data.obsolete.Label;
import org.metabrainz.mobile.api.data.obsolete.Recording;
import org.metabrainz.mobile.api.data.obsolete.Release;
import org.metabrainz.mobile.api.data.obsolete.ReleaseGroup;
import org.metabrainz.mobile.api.data.CoverArt;
import org.metabrainz.mobile.api.data.search.entity.Artist;

import java.io.IOException;

import io.reactivex.Single;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface LookupService {

    @GET("artist/{MBID}")
    Call<Artist> lookupArtist(@Path("MBID") String MBID,
                              @Query("inc") String params);

    @GET("http://coverartarchive.org/release/{MBID}")
    Single<CoverArt> getCoverArt(@Path("MBID") String MBID);

    @GET("https://en.wikipedia.org/api/rest_v1/page/summary/{title}")
    Call<ArtistWikiSummary> getWikipediaSummary(@Path("title") String title);

    @GET("https://www.wikidata.org/w/api.php" +
            "?action=wbgetentities&format=xml&props=sitelinks/urls&format=json")
    Call<ResponseBody> getWikipediaLink(@Query("ids") String id);

    @GET("release/{MBID}")
    Call<Release> lookupRelease(@Path("MBID") String MBID,
                                @Query("inc") String params);
    @GET("label/{MBID}")
    Call<Label> lookupLabel(@Path("MBID") String MBID,
                            @Query("inc") String params);
    @GET("recording/{MBID}")
    Call<Recording> lookupRecording(@Path("MBID") String MBID,
                                    @Query("inc") String params);
    @GET("release/{MBID}")
    Call<ReleaseGroup> lookupReleaseGroup(@Path("MBID") String MBID,
                                          @Query("inc") String params);

    // TODO: Implement Barcode LookUp
    Release lookupReleaseUsingBarcode(String barcode) throws IOException;
}

File: app/src/main/java/org/metabrainz/android/data/sources/api/LookupService.kt
package org.metabrainz.android.data.sources.api

import okhttp3.ResponseBody
import org.metabrainz.android.data.sources.api.entities.CoverArt
import org.metabrainz.android.data.sources.api.entities.RecordingItem
import org.metabrainz.android.data.sources.api.entities.WikiSummary
import org.metabrainz.android.data.sources.api.entities.acoustid.AcoustIDResponse
import org.metabrainz.android.data.sources.api.entities.mbentity.Release
import org.metabrainz.android.data.sources.api.entities.response.BarcodeReleaseResponse
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

interface LookupService {
    @GET("{entity}/{MBID}")
    suspend fun lookupEntityData(@Path("entity") entity: String,
                                 @Path("MBID") MBID: String,
                                 @Query("inc") params: String): ResponseBody

    @GET("http://coverartarchive.org/release/{MBID}")
    suspend fun getCoverArt(@Path("MBID") MBID: String?): CoverArt


    @GET("https://labs.api.listenbrainz.org/mbid-mapping/json")
    suspend fun searchRecording(@Query("[artist_credit_name]") artistCreditName: String?,
                        @Query("[recording_name]") recordingName: String?): ArrayList<RecordingItem>

    @GET("release/{MBID}")
    suspend fun lookupRecording(@Path("MBID") MBID: String?,
                        @Query("inc") params: String?): Release

    @POST("https://api.acoustid.org/v2/lookup")
    suspend fun lookupFingerprint(@Query("client") client: String?,
                          @Query("meta") metadata: String?,
                          @Query("duration") duration: Long,
                          @Query("fingerprint") fingerprint: String?): AcoustIDResponse

    @GET("https://en.wikipedia.org/api/rest_v1/page/summary/{title}")
    suspend fun getWikipediaSummary(@Path("title") title: String): WikiSummary

    @GET("https://www.wikidata.org/w/api.php" + "?action=wbgetentities&format=xml&props=sitelinks/urls&format=json")
    suspend fun getWikipediaLink(@Query("ids") id: String): ResponseBody

    @GET("release/")
    suspend fun lookupReleaseWithBarcode(@Query("query") barcode: String): BarcodeReleaseResponse
}